src/ray/util/shared_lru.h

```diff
@@ -0,0 +1,135 @@
+// // Get a key-value pair from cache.
+// auto val = ache.Get("key");
+// // Check and consume `val`.
+//
+// TODO(hjiang):
+// 1. Add template arguments for key hash and key equal, to pass into absl::flat_hash_map.
+// 2. Provide a key hash wrapper to save a copy.
+// 3. flat hash map supports heterogeneous lookup, expose `KeyLike` templated interface.
+
+#pragma once
+
+#include <cstdint>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <string>
+#include <utility>
+
+#include "absl/container/flat_hash_map.h"
+
+namespace ray::utils::container {
+
+template <typename Key, typename Val>
+class SharedLruCache final {
+ public:
+  using key_type = Key;
+  using mapped_type = Val;
+
+  // A `max_entries` of 0 means that there is no limit on the number of entries
+  // in the cache.
+  explicit SharedLruCache(size_t max_entries) : max_entries_(max_entries) {}
+
+  ~SharedLruCache() = default;
+
+  // Insert `value` with key `key`. This will replace any previous entry with
+  // the same key.
+  void Put(Key key, Val value) {
+    std::lock_guard lck(mu_);
+    lru_list_.emplace_front(key);
+    Entry new_entry{std::move(value), lru_list_.begin()};
+    cache_[key] = std::move(new_entry);
+
+    if (max_entries_ > 0 && lru_list_.size() > max_entries_) {
+      const auto &stale_key = lru_list_.back();
+      cache_.erase(stale_key);
+      lru_list_.pop_back();
+    }
+  }
+
+  // Delete the entry with key `key`. Return true if the entry was found for
+  // `key`, false if the entry was not found. In both cases, there is no entry
+  // with key `key` existed after the call.
+  bool Delete(const Key &key) {
+    std::lock_guard lck(mu_);
+    auto it = cache_.find(key);
+    if (it == cache_.end()) {
+      return false;
+    }
+    cache_.erase(it);
+    return true;
+  }
+
+  // Look up the entry with key `key`. Return std::nullopt if key doesn't exist.
+  // If found, return a copy for the value.
+  std::optional<Val> Get(const Key &key) {
+    std::lock_guard lck(mu_);
+    const auto cache_iter = cache_.find(key);
+    if (cache_iter == cache_.end()) {
+      return std::nullopt;
+    }
+    Val value = std::move(cache_iter->second.value);
+    lru_list_.erase(cache_iter->second.lru_iterator);
+    cache_.erase(cache_iter);
+
+    // Re-insert into the cache, no need to check capacity.
+    lru_list_.emplace_front(key);
+    Entry new_entry{value, lru_list_.begin()};
+    cache_[key] = std::move(new_entry);
+
+    return value;
+  }
+
+  // Clear the cache.
+  void Clear() {
+    std::lock_guard lck(mu_);
+    cache_.clear();
+    lru_list_.clear();
+  }
+
+  // Accessors for cache parameters.
+  size_t max_entries() const { return max_entries_; }
+
+ private:
+  struct Entry {
+    // The entry's value.
+    Val value;
+
+    // A list iterator pointing to the entry's position in the LRU list.
+    typename std::list<Key>::iterator lru_iterator;
+  };
+
+  using EntryMap = absl::flat_hash_map<Key, Entry>;
```

Comment:
how much do we expect this to be concurrently hit, haven't fully thought about it yet, but shouldn't have to lock for the whole time on every function, with concurrent_flat_map, and atomic iterators or lock free deques and stuff

Reply:
Fair point, I split the class into two classes.

Reply:
> shouldn't have to lock for the whole time on every function, with concurrent_flat_map, and atomic iterators or lock free deques and stuff

- First, I want to stress I'm not a fan of atomic operation, which is tricky to implement correct
- Second, we have multiple containers here, including list and hash map, don't think it's enough to have each of them thread-safe without a global lock

Reply:
ok, fine with it for now, can you leave a todo for me/you or both, to look into more perf thread-safe impl or getting boost lru cache to link with opencl

Reply:
> getting boost lru cache to link with opencl

boost lru cache is an internal util class, you could see it's under `details` namespace;
generally it's not recommended (but not completely disallowed) to use, if we have better alternatives.